presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java

```diff
@@ -0,0 +1,235 @@
+import static com.facebook.presto.cost.PlanNodeCostEstimate.cpuCost;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Scope.LOCAL;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Simple implementation of CostCalculator. It assumes that ExchangeNodes are already in the plan.
+ */
+@ThreadSafe
+public class CostCalculatorUsingExchanges
+        implements CostCalculator
+{
+    private final IntSupplier numberOfNodes;
+
+    @Inject
+    public CostCalculatorUsingExchanges(InternalNodeManager nodeManager)
+    {
+        this(() -> nodeManager.getAllNodes().getActiveNodes().size());
+    }
+
+    public CostCalculatorUsingExchanges(IntSupplier numberOfNodes)
+    {
+        this.numberOfNodes = requireNonNull(numberOfNodes, "numberOfNodes is null");
+    }
+
+    @Override
+    public PlanNodeCostEstimate calculateCost(PlanNode node, StatsProvider stats, Lookup lookup, Session session, Map<Symbol, Type> types)
+    {
+        CostEstimator costEstimator = new CostEstimator(numberOfNodes.getAsInt(), stats, lookup);
+        return node.accept(costEstimator, null);
+    }
+
+    private static class CostEstimator
+            extends PlanVisitor<PlanNodeCostEstimate, Void>
+    {
+        private final int numberOfNodes;
+        private final StatsProvider stats;
+        private final Lookup lookup;
+
+        CostEstimator(int numberOfNodes, StatsProvider stats, Lookup lookup)
+        {
+            this.numberOfNodes = numberOfNodes;
+            this.stats = requireNonNull(stats, "stats is null");
+            this.lookup = requireNonNull(lookup, "lookup is null");
+        }
+
+        @Override
+        protected PlanNodeCostEstimate visitPlan(PlanNode node, Void context)
+        {
+            return UNKNOWN_COST;
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitGroupReference(GroupReference node, Void context)
+        {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitOutput(OutputNode node, Void context)
+        {
+            return ZERO_COST;
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitFilter(FilterNode node, Void context)
+        {
+            return cpuCost(getStats(node.getSource()).getOutputSizeInBytes());
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitProject(ProjectNode node, Void context)
+        {
+            return cpuCost(getStats(node).getOutputSizeInBytes());
+        }
+
+        @Override
+        public PlanNodeCostEstimate visitAggregation(AggregationNode node, Void context)
+        {
+            PlanNodeStatsEstimate aggregationStats = getStats(node);
+            PlanNodeStatsEstimate sourceStats = getStats(node.getSource());
+            return new PlanNodeCostEstimate(sourceStats.getOutputSizeInBytes(), aggregationStats.getOutputSizeInBytes(), 0);
```

Comment:
using `cpuCost(...).memoryCost(...)` is much easier to read

Reply:
Easier, but i wouldn't say much. Since this is only in few places, I wouldn't add special support for that.
For greater readability, i added assignment to a variable before constructor call